lib/aws/rails/action_mailbox/sns_notification.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+require 'aws-sdk-sns'
+require 'aws-sdk-s3'
+
+module Aws
+  module Rails
+    module ActionMailbox
+      # @api private
+      class SnsNotification
+        class MessageContentError < StandardError; end
+
+        def initialize(request_body)
+          @request_body = request_body
+        end
+
+        def subscription_confirmed?
+          (200..299).cover?(confirmation_response.code.to_i)
+        end
+
+        def verified?
+          Aws::SNS::MessageVerifier.new.authentic?(@request_body)
+        end
+
+        def topic
+          notification.fetch(:TopicArn)
+        end
+
+        def type
+          notification.fetch(:Type)
+        end
+
+        def message_content
+          raise MessageContentError, 'Incoming emails must have notificationType `Received`' unless receipt?
+
+          if content_in_s3?
+            s3_content
+          else
+            return message[:content] unless destination
+
+            "X-Original-To: #{destination}\n#{message[:content]}"
+          end
+        end
+
+        private
+
+        def notification
+          @notification ||= JSON.parse(@request_body, symbolize_names: true)
+        rescue JSON::ParserError => e
+          Rails.logger.warn("Unable to parse SNS notification: #{e}")
+          nil
+        end
+
+        def s3_content
```

Comment:
We should memoize the client in initialize of this class or a higher level class (maybe in the controller somehow?) and pass it through.

Reply:
I am pretty sure if you memoize it in the controller, it's still regenerated each request, so it doesn't matter.

Maybe in an initializer somewhere?

Reply:
Yeah, in an initializer or railstie somehow and passed through. The client needs to be configurable through code.

Reply:
The only issue is that it'll change how the code works—right now region is pulled from the notification rather than being static.

Reply:
Why does it need to be configurable? For like a MVP release?

Reply:
We could make the region for the topic as config option then I can memoize it...it's another step, potentially, for setup, but it can be done.

Reply:
Okay, done....let me know if you like how I did it. I think it'll memoize properly.

Reply:
Looking at the example config in the README, it lists two queues in different regions, but configures the s3 client with only one region.  I think we may still have an issue with this memoization.  

Initializing clients is expensive (esp S3), so I agree we should be memoizing this somehow and agree that in general they should be configurable.  Maybe a method to retrieve/create an s3 client for a given region which would merge the `ses.s3_client_options`?

Something like:
```
def s3_client(region)
  @clients[region] ||= Aws::S3::Client.new(Rails.configuration.action_mailbox.ses.s3_client_options.merge(region: region)
end
```

For reference, the config I'm referring to is:
```
# config/environments/production.rb
config.action_mailbox.ses.subscribed_topics = %w(
  arn:aws:sns:eu-west-1:012345678910:example-topic-1
  arn:aws:sns:us-east-1:012345678910:example-topic-2
)
config.action_mailbox.ses.s3_client_options = { region: 'us-east-1' }
```

Reply:
How about we don't allow multiple subscribed topics for now and let that be an additional release (and someone can figure out how to easily do the s3 client stuff)?

I don't use multiple so I don't care 😁

Reply:
I think we can try to help with this.